lib/Parse/ParseDecl.cpp

```diff
@@ -799,6 +799,269 @@ Parser::parseImplementsAttribute(SourceLoc AtLoc, SourceLoc Loc) {
+      diagnose(Tok, diag::unexpected_separator, ",");
+      return true;
+    }
+    // Check that token after comma is 'wrt:' or a function specifier label.
+    if (!Tok.is(tok::identifier) || !(Tok.getText() == "wrt" ||
+                                      Tok.getText() == "jvp" ||
+                                      Tok.getText() == "vjp")) {
+      diagnose(Tok, diag::attr_differentiable_expected_label);
+      return true;
+    }
+    return false;
+  };
+
+  // Store starting parser position.
+  auto startingLoc = Tok.getLoc();
+  SyntaxParsingContext ContentContext(
+      SyntaxContext, SyntaxKind::DifferentiableAttributeArguments);
+
+  // Parse optional differentiation parameters.
+  // Parse 'linear' label (optional).
+  linear = false;
+  if (Tok.is(tok::identifier) && Tok.getText() == "linear") {
+    linear = true;
+    consumeToken(tok::identifier);
+    // If no trailing comma or 'where' clause, terminate parsing arguments.
+    if (Tok.isNot(tok::comma) && Tok.isNot(tok::kw_where))
+      return false;
+    if (consumeIfTrailingComma())
+      return errorAndSkipToEnd();
+  }
+
+  // If 'withRespectTo' is used, make the user change it to 'wrt'.
+  if (Tok.is(tok::identifier) && Tok.getText() == "withRespectTo") {
+    SourceRange withRespectToRange(Tok.getLoc(), peekToken().getLoc());
+    diagnose(Tok, diag::attr_differentiable_use_wrt_not_withrespectto)
+        .highlight(withRespectToRange)
+        .fixItReplace(withRespectToRange, "wrt:");
+    return errorAndSkipToEnd();
+  }
+  if (Tok.is(tok::identifier) && Tok.getText() == "wrt") {
+    if (parseDifferentiationParametersClause(params, AttrName))
+      return true;
+    // If no trailing comma or 'where' clause, terminate parsing arguments.
+    if (Tok.isNot(tok::comma) && Tok.isNot(tok::kw_where))
+      return false;
+    if (consumeIfTrailingComma())
+      return errorAndSkipToEnd();
+  }
+
+  // Function that parses a label and a function specifier, e.g. 'vjp: foo(_:)'.
+  // Return true on error.
+  auto parseFuncSpec = [&](StringRef label, DeclNameWithLoc &result,
+                           bool &terminateParsingArgs) -> bool {
+    // Parse label.
+    if (parseSpecificIdentifier(label,
+            diag::attr_differentiable_missing_label, label) ||
+        parseToken(tok::colon, diag::expected_colon_after_label, label))
+      return true;
+    // Parse the name of the function.
+    SyntaxParsingContext FuncDeclNameContext(
+         SyntaxContext, SyntaxKind::FunctionDeclName);
+    Diagnostic funcDiag(diag::attr_differentiable_expected_function_name.ID,
+                        { label });
+    result.Name =
+        parseUnqualifiedDeclName(/*afterDot=*/false, result.Loc,
+                                 funcDiag, /*allowOperators=*/true,
+                                 /*allowZeroArgCompoundNames=*/true);
+    // If no trailing comma or 'where' clause, terminate parsing arguments.
+    if (Tok.isNot(tok::comma) && Tok.isNot(tok::kw_where))
```

Comment:
`Tok.isNot` takes multiple arguments, please use that instead of `&&` here and anywhere else this comes up.